posts/inside-rust/2021-01-15-rustdoc-performance-improvements.md

```diff
@@ -0,0 +1,134 @@
+---
+layout: post
+title: "Rustdoc performance improvements"
+author: Joshua Nelson and Guillaume Gomez
+team: The Rustdoc Team <https://www.rust-lang.org/governance/teams/dev-tools#rustdoc>
+---
+
+Hi everyone! [**@GuillaumeGomez**] recently tweeted about the rustdoc performance improvements and suggested that we write a blog post about it:
+
+<blockquote class="twitter-tweet"><p lang="en" dir="ltr">The performance comparison between <a href="https://twitter.com/rustlang?ref_src=twsrc%5Etfw">@rustlang</a> rustdoc now and rustdoc 4 months ago are crazy! The rustdoc cleanup going on (we&#39;re still far from done!) is having a huge positive impact! Can&#39;t wait to see how much better the results will be.<br><br>Maybe I should write a blog post? <a href="https://t.co/XapdmdZ1IZ">pic.twitter.com/XapdmdZ1IZ</a></p>&mdash; Guillaume Gomez (@imperioworld_) <a href="https://twitter.com/imperioworld_/status/1349383125051305984?ref_src=twsrc%5Etfw">January 13, 2021</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 
+
+The tweet received a lot of comments approving the blog post idea so here we go!
+
+## It's all about cleanup
+
+With the recent growth of the rustdoc team, we finally had some time to pay the technical debt we've been accumulating for a while. To sum it up: removing implementations in rustdoc and use the compiler types directly. First, we need to explain a bit how rustdoc works. When we run it to generate HTML documentation, it goes through several steps:
+
+* Run parts of the compiler to get the information we need.
+* Remove the information provided by the compiler that we don't need (for example, if an item is "doc hidden", we don't need it). There is a lot to say on this part so maybe we'll write another blog post to go more into details.
+* The `doctree` pass which adds some extra information needed by rustdoc on some items of the compiler.
+* The `clean` pass which converts the compiler types into rustdoc ones: basically, it transforms everything into "printable" content.
+* The `render` pass which then generates the desired output (HTML or, on nightly, JSON).
+
+[**@jyn514**] noticed a while ago that [most of the work in Rustdoc is duplicated](https://github.com/rust-lang/rust/issues/76382): there are actually *three* different abstract syntax trees (ASTs)! One for `doctree`, one for `clean`, and one is the original [HIR](https://rustc-dev-guide.rust-lang.org/hir.html) used by the compiler.
+Rustdoc was spending quite a lot of time converting between them. Most of the speed improvements have come from getting rid of parts of the AST altogether.
+
+### Burning down the tree
```

Comment:
That sounds a bit harsh and not environmentally friendly... maybe
```suggestion
### Pruning the tree
```

Reply:
Is it relevant here? It's a pun and it's funny.

Reply:
How about a compromise: I change it to 'pruning the tree' if you let me add a footnote that it was changed to be environmentally friendly :joy: that just cracks me up.